tfhe/src/shortint/engine/server_side/mod.rs

```diff
@@ -205,34 +198,54 @@ impl ShortintEngine {
         programmable_bootstrap_lwe_ciphertext_mem_optimized(
             &ciphertext_buffers.buffer_lwe_after_ks,
             &mut ct.ct,
-            &ciphertext_buffers.accumulator,
+            &ciphertext_buffers.accumulator.acc,
             fourier_bsk,
             fft,
             stack,
         );
 
+        ct.degree = ciphertext_buffers.accumulator.degree;
+
         Ok(())
     }
 
-    pub(crate) fn keyswitch_programmable_bootstrap(
+    pub(crate) fn clear_carry<OpOrder: PBSOrderMarker>(
         &mut self,
         server_key: &ServerKey,
-        ct: &Ciphertext,
-        acc: &Accumulator,
-    ) -> EngineResult<Ciphertext> {
-        let mut ct_res = ct.clone();
-        self.keyswitch_programmable_bootstrap_assign(server_key, &mut ct_res, acc)?;
-        Ok(ct_res)
+        ct: &CiphertextBase<OpOrder>,
+    ) -> EngineResult<CiphertextBase<OpOrder>> {
+        let mut ct_in = ct.clone();
+        self.clear_carry_assign(server_key, &mut ct_in)?;
+        Ok(ct_in)
+    }
+
+    pub(crate) fn clear_carry_assign<OpOrder: PBSOrderMarker>(
+        &mut self,
+        server_key: &ServerKey,
+        ct: &mut CiphertextBase<OpOrder>,
+    ) -> EngineResult<()> {
+        match OpOrder::pbs_order() {
+            PBSOrder::KeyswitchBootstrap => {
+                let ct = unsafe { std::mem::transmute(ct) };
+                self.keyswitch_bootstrap_assign(server_key, ct)?;
+            }
+            PBSOrder::BootstrapKeyswitch => {
+                let ct = unsafe { std::mem::transmute(ct) };
+                self.bootstrap_keyswitch_assign(server_key, ct)?;
+            }
```

Comment:
If we are to do this unsafe transmutes, I think we must make the `PBSOrderMaker` "sealed" so that only
the 2 marker defined in this create can implement this trait.
https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed


Without that seal, user code could create its own marker struct that implement the `PBSOrderMarker`, make it return the wrong order and then we'd be transmuting from `CiphertextBase<Something>` to `CiphertextBig` / `CiphertextSmall` which to at best lead to incorrect result / a panic

Also the nomicon tells this about `transmute`
https://doc.rust-lang.org/nomicon/transmutes.html

> When transmuting between different compound types, you have to make sure they are laid out the same way! If layouts differ, the wrong fields are going to get filled with the wrong data, which will make you unhappy and can also be Undefined Behavior (see above).

> So how do you know if the layouts are the same? For repr(C) types and repr(transparent) types, layout is precisely defined. But for your run-of-the-mill repr(Rust), it is not. Even different instances of the same generic type can have wildly different layout. Vec<i32> and Vec<u32> might have their fields in the same order, or they might not. The details of what exactly is and is not guaranteed for data layout are still being worked out over [at the UCG WG](https://rust-lang.github.io/unsafe-code-guidelines/layout.html).

Reply:
Very good point, this trait is supposed to be sealed anyways